src/resampling.jl

```diff
@@ -129,3 +130,221 @@ Upsamples by a factor of two and applies the abs2 operation. The code is optimiz
+
+julia> b = resample_czt(a, (0.99,1.3)); # just zooming and shrinking
+
+julia> c = resample_czt(a, (x->0.8+x^2/3,x->1.0+0.2*x)); # a more complicated distortion
+
+julia> d = resample_czt(a, (x->1.0,x->1.0), shear=(x->50*x^2,0.0)); # a more complicated distortion
+
+julia> @ve a,b,c,d # visualize distortions
+```
+"""
+function resample_czt(arr::AbstractArray{T,N}, rel_zoom; shear=nothing, shear_dim=nothing, fix_nyquist=false, new_size = size(arr), rel_pad=0.2, do_damp=false, center=CtrMid) where {T,N}
+    RT = real(T)
+    orig_size = size(arr)
+    if do_damp
+        arr = damp_edge_outside(arr, rel_pad)
+    else
+        arr = copy(arr)
+    end
+    for d in 1:length(rel_zoom)
+        sd = mod(d,ndims(arr))+1
+        if !isnothing(shear_dim)
+            sd = shear_dim[d]
+        end
+        my_zoom = 1.0
+        # case of a list of zoom numbers
+        if (isa(rel_zoom, Tuple) && isa(rel_zoom[1], Number)) || isa(rel_zoom, Number)
+            my_zoom = rel_zoom[d]
+            f_res = ift(arr, d)
+            myshear = 0
+            if !isnothing(shear)
+                myshear =  shear[d]
+            end
+            if myshear != 0
+                shifts = ramp(real(eltype(f_res)),d, size(f_res,d), scale=ScaFT)
+                FourierTools.apply_shift_strength!(f_res, f_res, shifts, d, sd,myshear, fix_nyquist)
+            end
+            if T<:Real
+                f_res = real(FourierTools.czt_1d(f_res, my_zoom, d))
+            else
+                f_res = T.(FourierTools.czt_1d(f_res, my_zoom, d))
+            end
+            select_region!(f_res, arr) 
+        # case of position dependent zoom functions
+        elseif (isa(rel_zoom, Tuple) && isa(rel_zoom[1], Function))
+            pos = 0
+            p = 1
+            shifts = ramp(real(eltype(arr)),d, size(arr,d), scale=ScaFT)
+            # note that the slicing removes the dimension opposed to the version above.
+            for slice in eachslice(arr, dims=d)
+                myshear = let 
+                    if !isnothing(shear)
+                        if isa(shear[sd], Function)
+                            # position-dependent shear
+                            shear[sd](pos) 
+                        else
+                            shear[sd]
+                        end
+                    else
+                        zero(RT)
+                    end
+                end
+                my_zoom = rel_zoom[d](pos)
+                # one-d since a slice was selected
+                f_res = ift(slice,1) 
+                if myshear != 0
+                    FourierTools.apply_shift_strength!(f_res, f_res, shifts[p], 1, 1, myshear, fix_nyquist)                
+                end
+                f_res = let 
+                    if T<:Real
+                        real(FourierTools.czt_1d(f_res, my_zoom, 1))
+                    else
+                        T.(FourierTools.czt_1d(f_res, my_zoom, 1))
```

Comment:
That does a full allocation. Required? Why is it needed?

Reply:
again

Reply:
I don't know what you mean. I see 40 lines of code here. What exactly performs a "full" allocation?

Reply:
`   T.(FourierTools.czt_1d(f_res, my_zoom, 1))` why is the `T.` needed?